pkg/controllers/capacityreservation/expiration/controller.go

```diff
@@ -0,0 +1,135 @@
+	controllerruntime "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	karpv1 "sigs.k8s.io/karpenter/pkg/apis/v1"
+	"sigs.k8s.io/karpenter/pkg/cloudprovider"
+	"sigs.k8s.io/karpenter/pkg/operator/injection"
+	"sigs.k8s.io/karpenter/pkg/utils/nodeclaim"
+
+	ec2types "github.com/aws/aws-sdk-go-v2/service/ec2/types"
+
+	v1 "github.com/aws/karpenter-provider-aws/pkg/apis/v1"
+	"github.com/aws/karpenter-provider-aws/pkg/providers/capacityreservation"
+)
+
+// The capacityreservation.expiration controller handles pre-emptive termination for nodes in expiring capacity-block
+// reservations. These nodes should be terminated 40 minutes before the capacity-block's end time, 10 minutes before EC2
+// begins reclaiming the instances. This is inline with the interruption notification emitted by EC2.
+type Controller struct {
+	clk           clock.Clock
+	kubeClient    client.Client
+	cloudProvider cloudprovider.CloudProvider
+	crProvider    capacityreservation.Provider
+}
+
+func NewController(
+	clk clock.Clock,
+	kubeClient client.Client,
+	cloudProvider cloudprovider.CloudProvider,
+	capacityReservationProvider capacityreservation.Provider,
+) *Controller {
+	return &Controller{
+		clk:           clk,
+		kubeClient:    kubeClient,
+		cloudProvider: cloudProvider,
+		crProvider:    capacityReservationProvider,
+	}
+}
+
+func (c *Controller) Name() string {
+	return "capacityreservation.expiration"
+}
+
+func (c *Controller) Reconcile(ctx context.Context) (reconcile.Result, error) {
+	ctx = injection.WithControllerName(ctx, c.Name())
+	ncs, err := nodeclaim.ListManaged(ctx, c.kubeClient, c.cloudProvider)
+	if err != nil {
+		return reconcile.Result{}, fmt.Errorf("listing instance types, %w", err)
+	}
+	ec2CRs, err := c.crProvider.List(ctx, lo.FilterMap(ncs, func(nc *karpv1.NodeClaim, _ int) (v1.CapacityReservationSelectorTerm, bool) {
+		id, ok := nc.Labels[cloudprovider.ReservationIDLabel]
+		return v1.CapacityReservationSelectorTerm{ID: id}, ok
+	})...)
+	if err != nil {
+		return reconcile.Result{}, fmt.Errorf("getting capacity reservations, %w", err)
+	}
+	expiringCRs := sets.New(lo.FilterMap(ec2CRs, func(ec2CR *ec2types.CapacityReservation, _ int) (string, bool) {
+		cr, err := v1.CapacityReservationFromEC2(c.clk, ec2CR)
+		if err != nil {
+			log.FromContext(ctx).WithValues("capacity-reservation-id", *ec2CR.CapacityReservationId).Error(err, "failed to parse capacity reservation")
+			return "", false
+		}
+		return cr.ID, cr.State == v1.CapacityReservationStateExpiring
+	})...)
+	toDelete := lo.Filter(ncs, func(nc *karpv1.NodeClaim, _ int) bool {
+		id, ok := nc.Labels[cloudprovider.ReservationIDLabel]
+		if !ok {
+			return false
+		}
+		return expiringCRs.Has(id)
+	})
+	errs := make([]error, len(toDelete))
+	workqueue.ParallelizeUntil(ctx, 10, len(toDelete), func(i int) {
```

Comment:
This is inline with the current deletion parallelization?

Reply:
Yeah, this is more-or-less what we do for handling interruption events today